handlers/weniwebchat/weniwebchat.go

```diff
@@ -0,0 +1,259 @@
+}
+
+func (h *handler) receiveMsg(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.Event, error) {
+	payload := &miPayload{}
+	err := handlers.DecodeAndValidateJSON(payload, r)
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+	}
+
+	// check message type
+	if payload.Type != "message" || (payload.Message.Type != "text" && payload.Message.Type != "image" && payload.Message.Type != "video" && payload.Message.Type != "audio" && payload.Message.Type != "file" && payload.Message.Type != "location") {
+		return nil, handlers.WriteAndLogRequestIgnored(ctx, h, channel, w, r, "ignoring request, unknown message type")
+	}
+
+	// check empty content
+	if payload.Message.Text == "" && payload.Message.MediaURL == "" && (payload.Message.Latitude == "" || payload.Message.Longitude == "") {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, errors.New("blank message, media or location"))
+	}
+
+	// build urn
+	urn, err := urns.NewURNFromParts(urns.ExternalScheme, payload.From, "", "")
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+	}
+
+	// parse timestamp
+	ts, err := strconv.ParseInt(payload.Message.TimeStamp, 10, 64)
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, fmt.Errorf("invalid timestamp: %s", payload.Message.TimeStamp))
+	}
+
+	// parse medias
+	var mediaURL string
+	if payload.Message.Type == "location" {
+		mediaURL = fmt.Sprintf("geo:%s,%s", payload.Message.Latitude, payload.Message.Longitude)
+	} else if payload.Message.MediaURL != "" {
+		mediaURL = payload.Message.MediaURL
+		payload.Message.Text = payload.Message.Caption
+	}
+
+	// build message
+	date := time.Unix(ts, 0).UTC()
+	msg := h.Backend().NewIncomingMsg(channel, urn, payload.Message.Text).WithReceivedOn(date).WithContactName(payload.From)
+
+	if mediaURL != "" {
+		msg.WithAttachment(mediaURL)
+	}
+
+	return handlers.WriteMsgsAndResponse(ctx, h, []courier.Msg{msg}, w, r)
+}
+
+var timestamp = ""
+
+type moPayload struct {
+	Type    string    `json:"type" validate:"required"`
+	To      string    `json:"to"   validate:"required"`
+	From    string    `json:"from" validate:"required"`
+	Message moMessage `json:"message"`
+}
+
+type moMessage struct {
+	Type         string   `json:"type"      validate:"required"`
+	TimeStamp    string   `json:"timestamp" validate:"required"`
+	Text         string   `json:"text,omitempty"`
+	MediaURL     string   `json:"media_url,omitempty"`
+	Caption      string   `json:"caption,omitempty"`
+	Latitude     string   `json:"latitude,omitempty"`
+	Longitude    string   `json:"longitude,omitempty"`
+	QuickReplies []string `json:"quick_replies,omitempty"`
+}
+
+func (h *handler) SendMsg(ctx context.Context, msg courier.Msg) (courier.MsgStatus, error) {
+	start := time.Now()
+	status := h.Backend().NewMsgStatusForID(msg.Channel(), msg.ID(), courier.MsgSent)
```

Comment:
The status should start assuming it is an error until we are sure the message was sent.